xray/core/dataset.py

```diff
@@ -1805,5 +1805,84 @@ def apply_over_both(lhs_data_vars, rhs_data_vars, lhs_vars, rhs_vars):
         ds._variables.update(new_vars)
         return ds
 
+    def diff(self, dim, n=1, coord_new='upper'):
+        """Calculate the n-th order discrete difference along given axis.
+
+        Parameters
+        ----------
+        dim : str, optional
+            Dimension over which to calculate the finite difference.
+
+        n : int, optional
+            The number of times values are differenced.
+
+        coord_new : str, optional
```

Comment:
instead of `coord_new`, maybe `label` would be a more appropriate name? Writing `ds.diff('x', label='upper')` looks cleaner to me than `ds.diff('x', coord_new='upper')`